modules/automode.cpp

```diff
@@ -0,0 +1,693 @@
+
+    CAutoModeUser* FindUserByHost(const CString& sHostmask,
+                                const CString& sChannel = "") {
+        for (const auto& it : m_msUsers) {
+            CAutoModeUser* pUser = it.second;
+
+            if (pUser->HostMatches(sHostmask) &&
+                (sChannel.empty() || pUser->ChannelMatches(sChannel))) {
+                return pUser;
+            }
+        }
+
+        return nullptr;
+    }
+
+    bool CheckAutoMode(const CNick& Nick, CChan& Channel) {
+        CAutoModeUser* pUser =
+            FindUserByHost(Nick.GetHostMask(), Channel.GetName());
+
+        if (!pUser) {
+            return false;
+        }
+
+        if (pUser->GetUserKey().Equals("__NOKEY__")) {
+            PutIRC("MODE " + Channel.GetName() + " +" + pUser->GetMode() + " " + Nick.GetNick());
+        } else {
+            // then insert this nick into the queue, the timer does the rest
+            CString sNick = Nick.GetNick().AsLower();
+            if (m_msQueue.find(sNick) == m_msQueue.end()) {
+                m_msQueue[sNick] = "";
+            }
+        }
+
+        return true;
+    }
+
+    void DelUser(const CString& sUser) {
+        map<CString, CAutoModeUser*>::iterator it =
+            m_msUsers.find(sUser.AsLower());
+
+        if (it == m_msUsers.end()) {
+            PutModule(t_s("No such user"));
+            return;
+        }
+
+        delete it->second;
+        m_msUsers.erase(it);
+        PutModule(t_f("User {1} removed")(sUser));
+    }
+
+    CAutoModeUser* AddUser(const CString& sUser, CString& sMode, const CString& sKey,
+                         const CString& sHosts, const CString& sChans) {
+        if (m_msUsers.find(sUser) != m_msUsers.end()) {
+            PutModule(t_s("That user already exists"));
+            return nullptr;
+        }
+
+        CAutoModeUser* pUser = new CAutoModeUser(sUser, sMode, sKey, sHosts, sChans);
+        m_msUsers[sUser.AsLower()] = pUser;
+        PutModule(t_f("User {1} added with hostmask(s) {2}")(sUser, sHosts));
+        return pUser;
+    }
+
+    bool ChallengeRespond(const CNick& Nick, const CString& sChallenge) {
+        // Validate before responding - don't blindly trust everyone
+        bool bValid = false;
+        bool bMatchedHost = false;
+        CAutoModeUser* pUser = nullptr;
+
+        for (const auto& it : m_msUsers) {
+            pUser = it.second;
+
+            // First verify that the person who challenged us matches a user's
+            // host
+            if (pUser->HostMatches(Nick.GetHostMask())) {
+                const vector<CChan*>& Chans = GetNetwork()->GetChans();
+                bMatchedHost = true;
+
+                // Also verify that they are opped in at least one of the user's
+                // chans
+                for (CChan* pChan : Chans) {
+                    const CNick* pNick = pChan->FindNick(Nick.GetNick());
+
+                    if (pNick) {
+                        if (pNick->HasPerm(CChan::Op) &&
```

Comment:
I believe half op or +q should be enough for some modes?

Reply:
Half-Op can only set +v.